core/src/test/java/org/infinispan/distribution/TriangleExceptionDuringMarshallingTest.java

```diff
@@ -0,0 +1,220 @@
+      globalBuilder.serialization().marshaller(new JavaSerializationMarshaller());
+      globalBuilder.serialization().allowList()
+                   .addClasses(MagicKey.class, MarshallingExceptionGenerator.class);
+
+      ConfigurationBuilder cacheBuilder = new ConfigurationBuilder();
+      ControlledConsistentHashFactory<?> chf =
+            new ControlledConsistentHashFactory.Default(new int[][]{{0, 1}, {1, 2}, {2, 0}});
+      cacheBuilder.clustering().cacheMode(CacheMode.DIST_SYNC)
+                  .hash().numSegments(NUM_SEGMENTS).consistentHashFactory(chf);
+
+      createCluster(globalBuilder, cacheBuilder, 3);
+
+      // Make sure we're using the triangle algorithm
+      AsyncInterceptorChain asyncInterceptorChain = extractInterceptorChain(cache(0));
+      assertTrue(asyncInterceptorChain.containsInterceptorType(TriangleDistributionInterceptor.class));
+   }
+
+   public void testExceptionDuringMarshallingOnOriginator() {
+      // fail during the first serialization (i.e. on the originator)
+      Object value = MarshallingExceptionGenerator.failOnSerialization(0);
+
+      Cache<Object, Object> originCache = cache(0);
+      MagicKey primaryKey = new MagicKey("primary", cache(0));
+      expectException(MarshallingException.class, () -> originCache.put(primaryKey, value));
+      assertCleanFailure(originCache, primaryKey);
+
+      MagicKey nonOwnerKey = new MagicKey("non-owner", cache(1));
+      expectException(MarshallingException.class, () -> originCache.put(nonOwnerKey, value));
+      assertCleanFailure(originCache, nonOwnerKey);
+
+      MagicKey backupKey = new MagicKey("backup", cache(2));
+      expectException(MarshallingException.class, () -> originCache.put(backupKey, value));
+      assertCleanFailure(originCache, backupKey);
+   }
+
+   public void testExceptionDuringMarshallingOnRemote() {
+      // fail during the second serialization, i.e. remotely
+      Object value = MarshallingExceptionGenerator.failOnSerialization(1);
+
+      Cache<Object, Object> originCache = cache(0);
+      // does not fail when the originator is the primary, as there is no remote serialization
+      MagicKey primaryKey = new MagicKey("primary", cache(0));
+      originCache.put(primaryKey, value);
+      originCache.remove(primaryKey);
+
+      // fails when the originator is not an owner
+      MagicKey nonOwnerKey = new MagicKey("non-owner", cache(1));
+      expectException(RemoteException.class, MarshallingException.class, () -> originCache.put(nonOwnerKey, value));
+      assertCleanFailure(originCache, nonOwnerKey);
+
+      // fails when the originator is a backup
+      MagicKey backupKey = new MagicKey("backup", cache(2));
+      expectException(RemoteException.class, MarshallingException.class, () -> originCache.put(backupKey, value));
+      assertCleanFailure(originCache, backupKey);
+   }
+
+   @Test(enabled = false, description = "See ISPN-12770")
```

Comment:
couldn't this be as simpler as catching and storing in `BackupWriteCommand` any exception from `readFrom()`?
then, when `invokeAsync()` is invoked, you would return a `CompletableFuture` with the exception: `return CompletableFutures.completedExceptionFuture(unmarshallException);`

Reply:
Definitely not as simple as that, because the keys and values are read by the subclasses of `BackupWriteCommand` in their implementation of `readFrom()`. But I'll add a comment to ISPN-12770.